### PR TITLE
Prevent stale GatewayClassConfig

### LINKED ISSUE #88: GatewayClassConfig can become stale
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request. Searching for pre-existing feature requests helps us consolidate datapoints for identical requirements into a single place, thank you!
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.

---

### Overview of the Issue

It appears as though our dirty-checking for `GatewayClassConfig`s has a bug and our `GatewayClassConfig` values can become stale.

### Reproduction Steps

- Deploy the gateway controller via `helm` with a managed gateway class config.
- Update the helm deployment with a slightly different configuration for the managed class config (i.e. change the `serviceType` value)
- Deploy a gateway
- You should see the old class config being used and serialized onto the Gateway annotations, meaning we're holding a stale reference to the configuration

Note: restarting the controller and re-creating the gateway uses the latest class config.

### PR DESCRIPTION
**Pull request by nathancoleman**
Fixes #88 

Changes proposed in this PR:
- When a `GatewayClassConfig` changes, evict all `GatewayClasses` that reference it from the cache

How I've tested this PR:
See instructions in #88 

How I expect reviewers to test this PR:
Do the following on both main and this branch, observing that the config annotation on the `Gateway` is correct on this branch:
```shell
$ ./dev/run

# In a separate terminal
$ kubectl apply -f ./dev/config/k8s/

# Observe config in Gateway annotation
$ kubectl get gateway test-gateway -o yaml

# Add serviceType to GatewayClassConfig
$ kubectl edit gatewayclassconfig test-gateway-class-config

# Delete the gateway and recreate it, expecting that it will pick up the new serviceType in the config annotation
$ kubectl delete gateway test-gateway
$ kubectl apply -f ./dev/config/k8s/

# Observe config in Gateway annotation
# On main, the added serviceType will not be included at this stage.
# On this branch, the serviceType will be included.
$ kubectl get gateway test-gateway -o yaml

# Quit the ./dev/run process, switch branches, and then
$ ./dev/run -r

# Start back at the first command
```

Checklist:
- [x] Tests added
- [ ] CHANGELOG entry added 
  > HashiCorp engineers only, community PRs should not add a changelog entry.
  > Entries should use present tense (e.g. Add support for...)

### CODE REVIEW

**Comment by nathancoleman** [comment]
@mikemorris what should I be doing with the `CHANGELOG` here? It looks outdated, but I know you were looking into an automated process.

File: internal/k8s/controllers/gateway_class_config_controller.go [hunk 1]
```diff
@@ -62,6 +64,16 @@ func (r *GatewayClassConfigReconciler) Reconcile(ctx context.Context, req ctrl.R
 	}
 
 	// we're creating or updating
+
+	// evict any class that's referencing an updated config from cache
+	if using, err := r.Client.GatewayClassesUsingConfig(ctx, gcc); err == nil && len(using.Items) > 0 {
+		for _, gc := range using.Items {
+			if err := r.Manager.DeleteGatewayClass(ctx, gc.Name); err != nil {
+				logger.Warn("error evicting cached gateway class referencing config", "error", err)
```
Comment (nathancoleman) [inline internal/k8s/controllers/gateway_class_config_controller.go:72]: @mikemorris @andrewstucki this is best effort right now. Should we keep it that way, or should we be erroring out if we potentially have a stale `GatewayClassConfig` in the cache?

**Review by nathancoleman — COMMENTED** [review]
The e2e test runs seem to be intermittently timing out. I can't imagine that being caused by this change, but I'm not sure if it's something anyone else has seen since we merged those last week. @mikemorris @andrewstucki ?

**Comment by mikemorris** [comment]
> @mikemorris what should I be doing with the `CHANGELOG` here? It looks outdated, but I know you were looking into an automated process.

Started implementation in https://github.com/hashicorp/consul-api-gateway/pull/101 but shouldn't block this, we can always add an entry afterwards.

File: internal/k8s/controllers/gateway_class_config_controller.go (hunk 1 above)
Comment (andrewstucki) [reply internal/k8s/controllers/gateway_class_config_controller.go:72]: think we should be fine right now

**Review by andrewstucki — COMMENTED** [review]
I would love part of an e2e test that essentially leverages what we were seeing before -- think you can do something like create a `GatewayClassConfig`, create `GatewayClass` that leverages it, update the `GatewayClassConfig`, deploy a `Gateway` and verify that the serialized config on the deployed gateway matches the updated config. Think it shouldn't be too bad just to plumb into something existing.

**Comment by nathancoleman** [comment]
@andrewstucki added some e2e coverage for this fix. My additions are passing; however, one of the existing tests is not. Seems possibly related to what you're actively working on - could you take a look?

File: internal/commands/server/k8s_e2e_test.go [hunk 2]
```diff
@@ -880,6 +873,41 @@ func conditionInSync(conditions []meta.Condition) bool {
...
+	gw := &gateway.Gateway{
+		ObjectMeta: meta.ObjectMeta{
+			Name:      gatewayName,
+			Namespace: namespace,
+		},
+		Spec: gateway.GatewaySpec{
+			GatewayClassName: gateway.ObjectName(gc.Name),
+			Listeners: []gateway.Listener{{
+				Name:     "https",
+				Port:     gateway.PortNumber(443),
```
Comment (andrewstucki) [inline internal/commands/server/k8s_e2e_test.go:893]: @nathancoleman So this is why your tests are failing. When you extracted this helper you set all of the listeners to bind to port `443` -- the problem is that to actually route through the listener that's deployed in `kind` from the localhost you need to do all sorts of port-forwarding in the `kind` cluster itself. If you want to keep this extracted helper, then you'll need to parameterize it to take a port and then pass in `e2e.HTTPPort(ctx)` to the Gateway that you actually want to check the routing through, which I believe is currently on line 302.

**Review by andrewstucki — CHANGES_REQUESTED** [review]
You'll need to fix up the `createGateway` helper method you created to fix up the tests, but almost there!

File: internal/commands/server/k8s_e2e_test.go (hunk 2 above)
Comment (nathancoleman) [reply internal/commands/server/k8s_e2e_test.go]: Good catch!

**Review by andrewstucki — APPROVED** [review]

**Pull request merged**